File: 2022/rust/day4/src/main.rs
use std::ops::RangeInclusive;

fn get_data<'a>() -> Vec<&'a str> {
    let data = include_str!("./res/input.txt");
    data.split("\n").collect()
}

fn to_range(input: Vec<&str>) -> RangeInclusive<i32> {
    let start: i32 = input[0].parse::<i32>().unwrap();
    let end: i32 = input[1].parse::<i32>().unwrap();
    start..=end
}

fn convert_data(data: Vec<&str>) -> Vec<(RangeInclusive<i32>, RangeInclusive<i32>)> {
    let mut ret: Vec<RangeInclusive<i32>> = vec![];
    for line in data {
        let c_split: Vec<&str> = line.split(',').filter(|&x| x != "").collect();
        ret.extend(c_split.iter().map(|x| to_range(x.split("-").collect())));
    }
    ret.chunks_exact(2)
        .map(|chunk| (chunk[0].clone(), chunk[1].clone()))
        .collect()
}

fn encapsulates_another(left: &RangeInclusive<i32>, right: &RangeInclusive<i32>) -> bool {
    left.start() <= right.start() && left.end() >= right.end()
}

fn check_ranges(ranges: &(RangeInclusive<i32>, RangeInclusive<i32>)) -> bool {
    let (left, right) = ranges;
    encapsulates_another(&left, &right) || encapsulates_another(&right, &left)
}

fn main() {
    let data: Vec<&str> = get_data();
    let converted_data = convert_data(data);
    let result_list: Vec<bool> = converted_data
        .iter()
        .map(|ranges| check_ranges(ranges))
        .filter(|&x| x)
        .collect();
    println!("{}", result_list.len());
}

File: 2022/hs/day4/app/Main.hs
module Main where
import Data.List.Split (splitOn)

parse :: String -> [[Int]]
parse =  map(map read .concatMap (splitOn "-")) .map (splitOn ",") . lines

inRange :: [Int] -> Bool
inRange [l1,r1,l2,r2] =  minimum [r1,r2] >= maximum [l1,l2] 
inRange _ = False

rangeContains:: [Int] -> Bool 
rangeContains [l1,r1,l2,r2] =  l1 <= l2 && r1 >= r2
rangeContains _ = False

rangeEncapsulates :: [Int] -> Bool
rangeEncapsulates [l1,r1,l2,r2] =  rangeContains [l1,r1,l2,r2] || rangeContains [l2,r2,l1,r1] 
rangeEncapsulates _ = False

taskA :: [[Int]] -> Int
taskA =   length . filter id . map rangeEncapsulates 

taskB :: [[Int]] -> Int
taskB =   length . filter id . map inRange

main :: IO ()
main = do
  content <- readFile "./res/input.txt"
  print $  taskA . parse $ content
  print $  taskB . parse $ content
